server-spi/src/main/java/org/keycloak/models/credential/RecoveryAuthnCodesCredentialModel.java

```diff
@@ -0,0 +1,106 @@
+package org.keycloak.models.credential;
+
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.keycloak.credential.CredentialMetadata;
+import org.keycloak.credential.CredentialModel;
+import org.keycloak.models.credential.dto.RecoveryAuthnCodeRepresentation;
+import org.keycloak.models.credential.dto.RecoveryAuthnCodesCredentialData;
+import org.keycloak.models.credential.dto.RecoveryAuthnCodesSecretData;
+import org.keycloak.models.utils.RecoveryAuthnCodesUtils;
+import org.keycloak.util.JsonSerialization;
+
+import java.io.IOException;
+import java.util.List;
+
+public class RecoveryAuthnCodesCredentialModel extends CredentialModel {
+
+    public static final String TYPE = "recovery-authn-codes";
+
+    public static final String RECOVERY_CODES_NUMBER_USED = "recovery-codes-number-used";
+    public static final String RECOVERY_CODES_NUMBER_REMAINING = "recovery-codes-number-remaining";
+    public static final String RECOVERY_CODES_GENERATE_NEW_CODES = "recovery-codes-generate-new-codes";
+
+    private final RecoveryAuthnCodesCredentialData credentialData;
+    private final RecoveryAuthnCodesSecretData secretData;
+    private final CredentialMetadata credentialMetadata = null;
+
+    private RecoveryAuthnCodesCredentialModel(RecoveryAuthnCodesCredentialData credentialData,
+            RecoveryAuthnCodesSecretData secretData) {
+        this.credentialData = credentialData;
+        this.secretData = secretData;
+    }
+
+    public RecoveryAuthnCodeRepresentation getNextRecoveryAuthnCode() {
+        return this.secretData.getCodes().get(0);
```

Comment:
If getCodes() retruns an empty list, ArrayIndexOutOfBoundsException will be thrown. We could wrap the return in an `Optional` or simply return `null`.

Is the idea to require users always call `allCodesUsed()` before caling `getNextRecoveryAuthnCode()`?

Reply:
I'll use an optional. Good idea.